File: client/src/main/java/net/treset/minecraftlauncher/data/LauncherFiles.java
package net.treset.minecraftlauncher.data;

import javafx.util.Pair;
import net.treset.mc_version_loader.json.GenericJsonParsable;
import net.treset.mc_version_loader.launcher.*;
import net.treset.minecraftlauncher.LauncherApplication;
import net.treset.minecraftlauncher.util.exception.FileLoadException;
import net.treset.minecraftlauncher.util.file.LauncherFile;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class LauncherFiles {
    private static final Logger LOGGER = LogManager.getLogger(LauncherFiles.class);

    private LauncherManifest mainManifest;
    private LauncherDetails launcherDetails;
    private LauncherManifest gameDetailsManifest;
    private LauncherManifest modsManifest;
    private List<Pair<LauncherManifest, LauncherModsDetails>> modsComponents;
    private LauncherManifest savesManifest;
    private List<LauncherManifest> savesComponents;
    private LauncherManifest instanceManifest;
    private List<Pair<LauncherManifest, LauncherInstanceDetails>> instanceComponents;
    private LauncherManifest javaManifest;
    private List<LauncherManifest> javaComponents;
    private LauncherManifest optionsManifest;
    private List<LauncherManifest> optionsComponents;
    private LauncherManifest resourcepackManifest;
    private List<LauncherManifest> resourcepackComponents;
    private LauncherManifest versionManifest;
    private List<Pair<LauncherManifest, LauncherVersionDetails>> versionComponents;

    public LauncherFiles() throws FileLoadException {
        reloadMainManifest();
        reloadLauncherDetails();

        LOGGER.debug("Loaded launcher details");
    }

    public void reloadAll() throws FileLoadException {
        reloadMainManifest();
        reloadLauncherDetails();
        reloadGameDetailsManifest();
        reloadModsManifest();
        reloadModsComponents();
        reloadSavesManifest();
        reloadSavesComponents();
        reloadInstanceManifest();
        reloadInstanceComponents();
        reloadJavaManifest();
        reloadJavaComponents();
        reloadOptionsManifest();
        reloadOptionsComponents();
        reloadResourcepackManifest();
        reloadResourcepackComponents();
        reloadVersionManifest();
        reloadVersionComponents();
    }

    public LauncherManifest getMainManifest() {
        return mainManifest;
    }

    public void reloadMainManifest() throws FileLoadException {
        String versionFile;
        try {
            versionFile = LauncherFile.of(LauncherApplication.config.BASE_DIR, LauncherApplication.config.MANIFEST_FILE_NAME).readString();
        } catch (IOException e) {
            throw new FileLoadException("Unable to load launcher manifest: file error", e);
        }

        mainManifest = LauncherManifest.fromJson(versionFile);
        if(mainManifest == null || mainManifest.getType() != LauncherManifestType.LAUNCHER) {
            mainManifest = null;
            throw new FileLoadException("Unable to load launcher manifest: incorrect contents");
        }
        mainManifest.setDirectory(LauncherApplication.config.BASE_DIR.getAbsolutePath());
        LOGGER.debug("Loaded launcher manifest");
    }

    public LauncherDetails getLauncherDetails() {
        return launcherDetails;
    }

    public void reloadLauncherDetails() throws FileLoadException {
        if(getMainManifest() == null || getMainManifest().getDetails() == null) {
            throw new FileLoadException("Unable to load launcher details: invalid main file");
        }
        String detailsFile;
        try {
            detailsFile = LauncherFile.of(LauncherApplication.config.BASE_DIR, mainManifest.getDetails()).readString();
        } catch (IOException e) {
            throw new FileLoadException("Unable to load launcher details: file error", e);
        }

        this.launcherDetails = LauncherDetails.fromJson(detailsFile);
        if(this.launcherDetails == null || this.launcherDetails.getVersionDir() == null || this.launcherDetails.getVersionType() == null|| this.launcherDetails.getVersionComponentType() == null|| this.launcherDetails.getSavesType() == null || this.launcherDetails.getSavesComponentType() == null || this.launcherDetails.getResourcepacksType() == null|| this.launcherDetails.getResourcepacksComponentType() == null|| this.launcherDetails.getResourcepacksDir() == null || this.launcherDetails.getAssetsDir() == null|| this.launcherDetails.getGamedataDir() == null|| this.launcherDetails.getGamedataType() == null || this.launcherDetails.getInstancesDir() == null|| this.launcherDetails.getInstanceComponentType() == null|| this.launcherDetails.getInstancesType() == null || this.launcherDetails.getJavaComponentType() == null|| this.launcherDetails.getJavasDir() == null || this.launcherDetails.getJavasType() == null || this.launcherDetails.getLibrariesDir() == null|| this.launcherDetails.getModsComponentType() == null || this.launcherDetails.getModsType() == null || this.launcherDetails.getOptionsDir() == null || this.launcherDetails.getOptionsComponentType() == null || this.launcherDetails.getOptionsType() == null || this.launcherDetails.getSavesComponentType() == null|| this.launcherDetails.getSavesType() == null) {
            this.launcherDetails = null;
            throw new FileLoadException("Unable to load launcher details: incorrect contents");
        }
        LOGGER.debug("Loaded launcher details");
    }

    public LauncherManifest getGameDetailsManifest() {
        return gameDetailsManifest;
    }

    public void reloadGameDetailsManifest() throws FileLoadException {
        gameDetailsManifest = reloadManifest(LauncherFile.ofRelative(getLauncherDetails().getGamedataDir()), LauncherManifestType.GAME);
    }

    public LauncherManifest getModsManifest() {
        return modsManifest;
    }

    public void reloadModsManifest() throws FileLoadException {
        if(getGameDetailsManifest() == null || getGameDetailsManifest().getComponents() == null || getGameDetailsManifest().getComponents().size() != 2) {
            throw new FileLoadException("Unable to load mods manifest: invalid configuration");
        }
        modsManifest = reloadManifest(LauncherFile.ofRelative(getLauncherDetails().getGamedataDir()), getGameDetailsManifest().getComponents().get(0), LauncherManifestType.MODS);
    }

    public List<Pair<LauncherManifest, LauncherModsDetails>> getModsComponents() {
        return modsComponents;
    }

    public void reloadModsComponents() throws FileLoadException {
        modsComponents = reloadComponents(getModsManifest(), LauncherFile.ofRelative(getLauncherDetails().getGamedataDir()), LauncherManifestType.MODS_COMPONENT, LauncherModsDetails.class, LauncherFile.ofRelative(getLauncherDetails().getGamedataDir(), "mods"));
    }

    public LauncherManifest getSavesManifest() {
        return savesManifest;
    }

    public void reloadSavesManifest() throws FileLoadException {
        savesManifest = reloadManifest(LauncherFile.ofRelative(getLauncherDetails().getGamedataDir()), getGameDetailsManifest().getComponents().get(1), LauncherManifestType.SAVES);
    }

    public List<LauncherManifest> getSavesComponents() {
        return savesComponents;
    }

    public void reloadSavesComponents() throws FileLoadException {
        savesComponents = reloadComponents(getSavesManifest(), LauncherFile.ofRelative(getLauncherDetails().getGamedataDir()), LauncherManifestType.SAVES_COMPONENT, LauncherFile.ofRelative(getLauncherDetails().getGamedataDir(), "saves"));
    }

    public LauncherManifest getInstanceManifest() {
        return instanceManifest;
    }

    public void reloadInstanceManifest() throws FileLoadException {
        instanceManifest = reloadManifest(LauncherFile.ofRelative(getLauncherDetails().getInstancesDir()), LauncherManifestType.INSTANCES);
    }

    public List<Pair<LauncherManifest, LauncherInstanceDetails>> getInstanceComponents() {
        return instanceComponents;
    }

    public void reloadInstanceComponents() throws FileLoadException {
        instanceComponents = reloadComponents(getInstanceManifest(), LauncherFile.ofRelative(getLauncherDetails().getInstancesDir()), LauncherManifestType.INSTANCE_COMPONENT, LauncherInstanceDetails.class, null);
    }


    public LauncherManifest getJavaManifest() {
        return javaManifest;
    }

    public void reloadJavaManifest() throws FileLoadException {
        javaManifest = reloadManifest(LauncherFile.ofRelative(getLauncherDetails().getJavasDir()), LauncherManifestType.JAVAS);
    }

    public List<LauncherManifest> getJavaComponents() {
        return javaComponents;
    }

    public void reloadJavaComponents() throws FileLoadException {
        javaComponents = reloadComponents(getJavaManifest(), LauncherFile.ofRelative(getLauncherDetails().getJavasDir()), LauncherManifestType.JAVA_COMPONENT, null);
    }

    public LauncherManifest getOptionsManifest() {
        return optionsManifest;
    }

    public void reloadOptionsManifest() throws FileLoadException {
        optionsManifest = reloadManifest(LauncherFile.ofRelative(getLauncherDetails().getOptionsDir()), LauncherManifestType.OPTIONS);
    }

    public List<LauncherManifest> getOptionsComponents() {
        return optionsComponents;
    }

    public void reloadOptionsComponents() throws FileLoadException {
        optionsComponents = reloadComponents(getOptionsManifest(), LauncherFile.ofRelative(getLauncherDetails().getOptionsDir()), LauncherManifestType.OPTIONS_COMPONENT, null);
    }

    public LauncherManifest getResourcepackManifest() {
        return resourcepackManifest;
    }

    public void reloadResourcepackManifest() throws FileLoadException {
        resourcepackManifest = reloadManifest(LauncherFile.ofRelative(getLauncherDetails().getResourcepacksDir()), LauncherManifestType.RESOURCEPACKS);
    }

    public List<LauncherManifest> getResourcepackComponents() {
        return resourcepackComponents;
    }

    public void reloadResourcepackComponents() throws FileLoadException {
        resourcepackComponents = reloadComponents(getResourcepackManifest(), LauncherFile.ofRelative(getLauncherDetails().getResourcepacksDir()), LauncherManifestType.RESOURCEPACKS_COMPONENT, null);
    }

    public LauncherManifest getVersionManifest() {
        return versionManifest;
    }

    public void reloadVersionManifest() throws FileLoadException {
        versionManifest = reloadManifest(LauncherFile.ofRelative(getLauncherDetails().getVersionDir()), LauncherManifestType.VERSIONS);
    }

    public List<Pair<LauncherManifest, LauncherVersionDetails>> getVersionComponents() {
        return versionComponents;
    }

    public void reloadVersionComponents() throws FileLoadException {
        versionComponents = reloadComponents(getVersionManifest(), LauncherFile.ofRelative(getLauncherDetails().getVersionDir()), LauncherManifestType.VERSION_COMPONENT, LauncherVersionDetails.class, null);
    }

    public LauncherManifest reloadManifest(LauncherFile path, LauncherManifestType expectedType) throws FileLoadException {
        return reloadManifest(path, LauncherApplication.config.MANIFEST_FILE_NAME, expectedType);
    }

    public LauncherManifest reloadManifest(LauncherFile path, String filename, LauncherManifestType expectedType) throws FileLoadException {
        String versionFile;
        try {
            versionFile = LauncherFile.of(path, filename).readString();
        } catch (IOException e) {
            throw new FileLoadException("Unable to load " + expectedType.name().toLowerCase() + " manifest: file error", e);
        }
        if(versionFile == null) {
            throw new FileLoadException("Unable to load " + expectedType.name().toLowerCase() + " manifest: file error");
        }
        LauncherManifest out = LauncherManifest.fromJson(versionFile, getLauncherDetails().getTypeConversion());
        if(out == null || out.getType() != expectedType) {
            throw new FileLoadException("Unable to load " + expectedType.name().toLowerCase() + " manifest: incorrect contents");
        }
        out.setDirectory(path.getAbsolutePath());
        LOGGER.debug("Loaded " + expectedType.name().toLowerCase() + " manifest");
        return out;
    }

    public List<LauncherManifest> reloadComponents(LauncherManifest parentManifest, LauncherFile parentPath, LauncherManifestType expectedType, LauncherFile fallbackPath) throws FileLoadException {
        return reloadComponents(parentManifest, parentPath, LauncherApplication.config.MANIFEST_FILE_NAME, expectedType, fallbackPath);
    }

    public List<LauncherManifest> reloadComponents(LauncherManifest parentManifest, LauncherFile parentPath, String filename, LauncherManifestType expectedType, LauncherFile fallbackPath) throws FileLoadException {
        List<LauncherManifest> out = new ArrayList<>();
        for(String c : parentManifest.getComponents()) {
            try {
                addComponent(out, LauncherFile.of(parentPath, parentManifest.getPrefix() + "_" + c), filename, expectedType, c, fallbackPath);
            } catch (FileLoadException e) {
                throw new FileLoadException("Unable to load " + expectedType.name().toLowerCase() + " components: component error: id=" + c);
            }
        }
        LOGGER.debug("Loaded " + expectedType.name().toLowerCase() + " components");
        return out;
    }

    private void addComponent(List<LauncherManifest> list, LauncherFile path, String filename, LauncherManifestType expectedType, String expectedId, LauncherFile fallbackPath) throws FileLoadException {
        String manifestFile;
        try {
            manifestFile = LauncherFile.of(path, filename).readString();
        } catch (IOException e) {
            if(fallbackPath == null) {
                throw new FileLoadException("Unable to load " + expectedType.name().toLowerCase() + " component: file error: id=" + expectedId, e);
            }
            LOGGER.debug("Falling back to fallback path loading " + expectedType.name().toLowerCase() + " component: file error: id=" + expectedId);
            addComponent(list, fallbackPath, filename, expectedType, expectedId, null);
            return;
        }

        LauncherManifest manifest = LauncherManifest.fromJson(manifestFile, getLauncherDetails().getTypeConversion());
        if(manifest == null || manifest.getType() == null || manifest.getType() != expectedType || manifest.getId() == null || !manifest.getId().equals(expectedId)) {
            if(fallbackPath == null) {
                throw new FileLoadException("Unable to load " + expectedType.name().toLowerCase() + " component: incorrect contents: id=" + expectedId);
            }
            LOGGER.debug("Falling back to fallback path loading " +  expectedType.name().toLowerCase() + " component id=" + expectedId);
            addComponent(list, fallbackPath, filename, expectedType, expectedId, null);
            return;
        }
        manifest.setDirectory(path.getAbsolutePath());
        list.add(manifest);
    }

    public <T extends GenericJsonParsable> List<Pair<LauncherManifest, T>> reloadComponents(LauncherManifest parentManifest, LauncherFile parentDir, LauncherManifestType expectedType, Class<T> targetClass, LauncherFile fallbackPath) throws FileLoadException {
        return reloadComponents(parentManifest, parentDir, LauncherApplication.config.MANIFEST_FILE_NAME, expectedType, targetClass, fallbackPath);
    }

    public <T extends GenericJsonParsable> List<Pair<LauncherManifest, T>> reloadComponents(LauncherManifest parentManifest, LauncherFile parentPath, String filename, LauncherManifestType expectedType, Class<T> targetClass, LauncherFile fallbackPath) throws FileLoadException {
        if(parentManifest == null || parentManifest.getPrefix() == null || parentManifest.getComponents() == null) {
            throw new FileLoadException("Unable to load " + expectedType.name().toLowerCase() + " components: invalid configuration");
        }
        List<Pair<LauncherManifest, T>> out = new ArrayList<>();
        List<FileLoadException> exceptionQueue = new ArrayList<>();
        for(String c : parentManifest.getComponents()) {
            try {
                addComponent(out, LauncherFile.of(parentPath, parentManifest.getPrefix() + "_" + c), filename, expectedType, targetClass, fallbackPath, c);
            } catch (FileLoadException e) {
                exceptionQueue.add(e);
            }
        }
        if(!exceptionQueue.isEmpty()) {
            throw new FileLoadException("Unable to load " + exceptionQueue.size() + " components: component error", exceptionQueue.get(0));
        }
        LOGGER.debug("Loaded " + expectedType.name().toLowerCase() + " components");
        return out;
    }


    private <T extends GenericJsonParsable> void addComponent(List<Pair<LauncherManifest, T>> list, LauncherFile path, String filename, LauncherManifestType expectedType, Class<T> targetClass, LauncherFile fallbackPath, String expectedId) throws FileLoadException {
        String manifestFile;
        try {
            manifestFile = LauncherFile.of(path, filename).readString();
        } catch (IOException e) {
            if(fallbackPath == null) {
                throw new FileLoadException("Unable to load " + expectedType.name().toLowerCase() + " component: file error: id=" + expectedId, e);
            }
            LOGGER.debug("Falling back to fallback path loading " + expectedType.name().toLowerCase() + " component: file error: id=" + expectedId);
            addComponent(list, fallbackPath, filename, expectedType, targetClass, null, expectedId);
            return;
        }

        LauncherManifest manifest = LauncherManifest.fromJson(manifestFile, getLauncherDetails().getTypeConversion());
        if(manifest == null || manifest.getType() == null || manifest.getType() != expectedType || manifest.getId() == null || !manifest.getId().equals(expectedId) || manifest.getDetails() == null) {
            if(fallbackPath == null) {
                throw new FileLoadException("Unable to load " + expectedType.name().toLowerCase() + " component: incorrect contents: id=" + expectedId);
            }
            LOGGER.debug("Falling back to fallback path loading " +  expectedType.name().toLowerCase() + " component id=" + expectedId);
            addComponent(list, fallbackPath, filename, expectedType, targetClass, null, expectedId);
            return;
        }
        manifest.setDirectory(path.getAbsolutePath());
        String detailsFile = null;
        try {
            detailsFile = LauncherFile.of(path, manifest.getDetails()).readString();
        } catch (IOException e) {
            if(fallbackPath == null) {
                throw new FileLoadException("Unable to load " + expectedType.name().toLowerCase() + " component details: file error: id=" + expectedId, e);
            }
            LOGGER.debug("Falling back to fallback path loading " +  expectedType.name().toLowerCase() + " component id=" + expectedId);
            addComponent(list, fallbackPath, filename, expectedType, targetClass, null, expectedId);
        }
        T details = GenericJsonParsable.fromJson(detailsFile, targetClass);
        if(details == null) {
            if(fallbackPath == null) {
                throw new FileLoadException("Unable to load " + expectedType.name().toLowerCase() + " component details: incorrect contents: id=" + expectedId);
            }
            LOGGER.debug("Falling back to fallback path loading " +  expectedType.name().toLowerCase() + " component id=" + expectedId);
            addComponent(list, fallbackPath, filename, expectedType, targetClass, null, expectedId);
            return;
        }
        list.add(new Pair<>(manifest, details));
    }
}

File: client/composeApp/src/commonMain/kotlin/net/treset/treelauncher/backend/data/LauncherFiles.kt
package net.treset.treelauncher.backend.data

import io.github.oshai.kotlinlogging.KotlinLogging
import net.treset.treelauncher.backend.config.appConfig
import net.treset.treelauncher.backend.data.manifest.*
import net.treset.treelauncher.backend.util.exception.FileLoadException
import net.treset.treelauncher.backend.util.file.LauncherFile
import java.io.FileNotFoundException
import java.io.IOException
import java.util.*

open class LauncherFiles {
    private var _mainManifest: MainManifest? = null
    val mainManifest: MainManifest
        get() = _mainManifest!!
    protected var _modsManifest: ParentManifest? = null
    val modsManifest: ParentManifest
        get() = _modsManifest!!
    protected var _modsComponents: Array<ModsComponent>? = null
    val modsComponents: Array<ModsComponent>
        get() = _modsComponents!!
    protected var _savesManifest: ParentManifest? = null
    val savesManifest: ParentManifest
        get() = _savesManifest!!
    protected var _savesComponents: Array<SavesComponent>? = null
    val savesComponents: Array<SavesComponent>
        get() = _savesComponents!!
    private var _instanceManifest: ParentManifest? = null
    val instanceManifest: ParentManifest
        get() = _instanceManifest!!
    private var _instanceComponents: Array<InstanceComponent>? = null
    val instanceComponents: Array<InstanceComponent>
        get() = _instanceComponents!!
    private var _javaManifest: ParentManifest? = null
    val javaManifest: ParentManifest
        get() = _javaManifest!!
    private var _javaComponents: Array<JavaComponent>? = null
    val javaComponents: Array<JavaComponent>
        get() = _javaComponents!!
    private var _optionsManifest: ParentManifest? = null
    val optionsManifest: ParentManifest
        get() = _optionsManifest!!
    private var _optionsComponents: Array<OptionsComponent>? = null
    val optionsComponents: Array<OptionsComponent>
        get() = _optionsComponents!!
    private var _resourcepackManifest: ParentManifest? = null
    val resourcepackManifest: ParentManifest
        get() = _resourcepackManifest!!
    private var _resourcepackComponents: Array<ResourcepackComponent>? = null
    val resourcepackComponents: Array<ResourcepackComponent>
        get() = _resourcepackComponents!!
    private var _versionManifest: ParentManifest? = null
    val versionManifest: ParentManifest
        get() = _versionManifest!!
    private var _versionComponents: Array<VersionComponent>? = null
    val versionComponents: Array<VersionComponent>
        get() = _versionComponents!!

    private var _gameDataDir: LauncherFile? = null
    val gameDataDir: LauncherFile
        get() = _gameDataDir!!

    init {
        reloadMain()
        LOGGER.debug { "Loaded Main manifest" }
    }

    @Throws(IOException::class)
    fun reload() {
        LOGGER.debug { "Reloading all components..." }
        reloadMain()
        reloadMods()
        reloadSaves()
        reloadInstances()
        reloadJavas()
        reloadOptions()
        reloadResourcepacks()
        reloadVersions()
        LOGGER.debug { "Finished reloading all components" }
    }

    @Throws(IOException::class)
    fun reloadMain() {
        LOGGER.debug { "Reloading main manifest..." }
        val file = LauncherFile.ofData(appConfig().manifestFileName)
        _mainManifest = MainManifest.readFile(file)
        _gameDataDir = LauncherFile.ofData(_mainManifest?.gameDataDir ?: throw FileLoadException("Unable to load main manifest: invalid configuration"))
        LOGGER.debug { "Finished reloading main manifest" }
    }


    @Throws(IOException::class)
    open fun reloadMods() {
        LOGGER.debug { "Reloading mods..." }
        val file = LauncherFile.ofData(_mainManifest?.modsDir ?: throw IOException("Unable to load mods manifest: invalid configuration"))
        _modsManifest = ParentManifest.readFile(file, LauncherManifestType.MODS)

        _modsComponents = reloadComponents(
            _modsManifest ?: throw FileNotFoundException("Unable to load mods components: parent manifest not loaded"),
            ModsComponent::readFile,
            _modsComponents ?: emptyArray()
        )
        LOGGER.debug { "Finished reloading mods" }
    }

    @Throws(IOException::class)
    open fun reloadSaves() {
        LOGGER.debug { "Reloading saves..." }
        val file = LauncherFile.ofData(_mainManifest?.savesDir ?: throw FileLoadException("Unable to load saves manifest: invalid configuration"))
        _savesManifest = ParentManifest.readFile(file, LauncherManifestType.SAVES)

        _savesComponents = reloadComponents(
            _savesManifest ?: throw FileNotFoundException("Unable to load saves components: parent manifest not loaded"),
            SavesComponent::readFile,
            _savesComponents?: emptyArray()
        )
        LOGGER.debug { "Finished reloading saves" }
    }

    @Throws(IOException::class)
    fun reloadInstances() {
        LOGGER.debug { "Reloading instances..." }
        val file = LauncherFile.ofData(_mainManifest?.instancesDir ?: throw FileLoadException("Unable to load instance manifest: invalid configuration"))
        _instanceManifest = ParentManifest.readFile(file, LauncherManifestType.INSTANCES)

        _instanceComponents = reloadComponents(
            _instanceManifest ?: throw FileNotFoundException("Unable to load instance components: parent manifest not loaded"),
            InstanceComponent::readFile,
            _instanceComponents?: emptyArray()
        )
        LOGGER.debug { "Finished reloading instances" }
    }

    @Throws(IOException::class)
    fun reloadJavas() {
        LOGGER.debug { "Reloading javas..." }
        val file = LauncherFile.ofData(_mainManifest?.javasDir ?: throw FileLoadException("Unable to load java manifest: invalid configuration"))
        _javaManifest = ParentManifest.readFile(file, LauncherManifestType.JAVAS)

        _javaComponents = reloadComponents(
            _javaManifest ?: throw FileNotFoundException("Unable to load java components: parent manifest not loaded"),
            JavaComponent::readFile,
            _javaComponents?: emptyArray()
        )
        LOGGER.debug { "Finished reloading javas" }
    }

    @Throws(IOException::class)
    fun reloadOptions() {
        LOGGER.debug { "Reloading options..." }
        val file = LauncherFile.ofData(_mainManifest?.optionsDir ?: throw FileLoadException("Unable to load options manifest: invalid configuration"))
        _optionsManifest = ParentManifest.readFile(file, LauncherManifestType.OPTIONS)

        _optionsComponents = reloadComponents(
            _optionsManifest ?: throw FileNotFoundException("Unable to load options components: parent manifest not loaded"),
            OptionsComponent::readFile,
            _optionsComponents?: emptyArray()
        )
        LOGGER.debug { "Finished reloading options" }
    }

    @Throws(IOException::class)
    fun reloadResourcepacks() {
        LOGGER.debug { "Reloading resourcepacks..." }
        val file = LauncherFile.ofData(_mainManifest?.resourcepacksDir ?: throw FileLoadException("Unable to load resourcepack manifest: invalid configuration"))
        _resourcepackManifest = ParentManifest.readFile(file, LauncherManifestType.RESOURCEPACKS)

        _resourcepackComponents = reloadComponents(
            _resourcepackManifest ?: throw FileNotFoundException("Unable to load resourcepack components: parent manifest not loaded"),
            ResourcepackComponent::readFile,
            _resourcepackComponents?: emptyArray()
        )
        LOGGER.debug { "Finished reloading resourcepacks" }
    }
    @Throws(IOException::class)
    fun reloadVersions() {
        LOGGER.debug { "Reloading versions..." }
        val file = LauncherFile.ofData(_mainManifest?.versionDir ?: throw FileLoadException("Unable to load version manifest: invalid configuration"))
        _versionManifest = ParentManifest.readFile(file, LauncherManifestType.VERSIONS)

        _versionComponents = reloadComponents(
            _versionManifest ?: throw FileNotFoundException("Unable to load version components: parent manifest not loaded"),
            VersionComponent::readFile,
            _versionComponents?: emptyArray()
        )
        LOGGER.debug { "Finished reloading versions" }
    }

    @Throws(IOException::class)
    inline fun <reified T: Component> reloadComponents(
        parentManifest: ParentManifest,
        load: (LauncherFile) -> T,
        currentComponents: Array<T>
    ): Array<T> {
        val out = mutableListOf<T>()

        for(id in parentManifest.components) {
            val file = LauncherFile.of(parentManifest.directory, "${parentManifest.prefix}_$id", appConfig().manifestFileName)
            val component = load(file)
            currentComponents.filter { it.id == component.id }.firstOrNull() ?.let {
                component.copyTo(it)
                out.add(it)
            } ?: out.add(component)
        }

        return out.toTypedArray()
    }

    @Throws(IOException::class)
    fun cleanupVersions(
        includeLibraries: Boolean
    ) {
        reload()

        LOGGER.debug { "Cleaning up versions..." }
        LOGGER.debug { "Checking for used versions..." }
        val usedVersions: MutableList<String> = mutableListOf()
        for (instance in instanceComponents) {
            usedVersions.add(instance.versionComponent)
            LOGGER.debug { "Used version: ${instance.versionComponent}" }
        }

        var newFound: Boolean
        do {
            LOGGER.debug { "Checking for dependencies..." }
            newFound = false
            for (version in versionComponents) {
                if (usedVersions.contains(version.id)) {
                    version.depends?.let {
                        if (!usedVersions.contains(it)) {
                            usedVersions.add(it)
                            newFound = true
                            LOGGER.debug { "Used version: ${version.id}" }
                        }
                    }
                }
            }
        } while(newFound)
        LOGGER.debug { "Finished checking for dependencies" }
        LOGGER.debug { "Finished checking for used versions" }

        LOGGER.debug { "Deleting unused versions..."}
        for (version in versionComponents) {
            if (!usedVersions.contains(version.id)) {
                LOGGER.debug { "Deleting unused version: ${version.id}" }
                versionManifest.components.remove(version.id)
                try {
                    LauncherFile.of(version.directory).remove()
                } catch(e: IOException) {
                    throw FileLoadException("Unable to remove unused version: file error", e)
                }
            }
        }
        LOGGER.debug { "Finished deleting unused versions" }

        LOGGER.debug { "Saving instance manifest..." }
        try {
            LauncherFile.of(
                versionManifest.directory,
                appConfig().manifestFileName
            ).write(versionManifest)
        } catch(e: IOException) {
            throw FileLoadException("Unable to save instance manifest: file error", e)
        }
        LOGGER.debug { "Saved instance manifest" }

        if(includeLibraries) {
            cleanupLibraries(usedVersions)
        }

        LOGGER.debug { "Finished cleaning up versions" }
    }

    private fun cleanupLibraries(
        usedVersions: List<String>
    ) {
        LOGGER.debug { "Cleaning up libraries..." }

        LOGGER.debug { "Collecting libraries in unused versions..." }
        val unusedLibraries: MutableList<String> = mutableListOf()
        for(version in versionComponents) {
            if(!usedVersions.contains(version.id)) {
                unusedLibraries.addAll(version.libraries)
                LOGGER.debug { "Unused libraries: ${version.libraries}" }
            }
        }
        LOGGER.debug { "Finished collecting libraries in unused versions" }

        LOGGER.debug { "Checking unused libraries against used versions..." }
        for(version in versionComponents) {
            if(usedVersions.contains(version.id)) {
                val libsToRemove = unusedLibraries.filter { version.libraries.contains(it) }
                if(libsToRemove.isNotEmpty()) {
                    unusedLibraries.removeAll(libsToRemove)
                    LOGGER.debug { "Removing used libraries from unused list: $libsToRemove" }
                }
            }
        }
        LOGGER.debug { "Finished checking unused libraries against used versions" }

        LOGGER.debug { "Deleting unused libraries..." }
        for(library in unusedLibraries) {
            val libFile = LauncherFile.of(versionManifest.directory, library)
            if(libFile.exists()) {
                try {
                    libFile.remove()
                } catch(e: IOException) {
                    LOGGER.warn(e) { "Unable to remove unused library, ignoring" }
                }
            }
        }
        LOGGER.debug { "Finished deleting unused libraries" }
        LOGGER.debug { "Finished cleaning up libraries" }
    }

    companion object {
        private val LOGGER = KotlinLogging.logger{}
    }
}

@Throws(IOException::class)
fun Array<out Component>.getId(id: String): Component {
    return this.firstOrNull { it.id == id } ?: throw FileNotFoundException("Unable to find component: id=$id")
}
